src/CarbonAware.Aggregators/src/CarbonAware/CarbonAwareParameters.cs

```diff
@@ -0,0 +1,215 @@
+            }
+        }
+        return mapping;
+    }
+}
+
+public class CarbonAwareParameters
+{
+    private IEnumerable<Location> _mulitpleLocations = new List<Location>();
+    private DateTimeOffset _start = DateTimeOffset.MinValue;
+    private DateTimeOffset _end = DateTimeOffset.MaxValue;
+
+    public enum PropertyName { MultipleLocations, Start, End };
+    public IEnumerable<Location> MultipleLocations {
+        get { return _mulitpleLocations; }
+        set { 
+            if (value.Any()) // Only set if non-empty
+            {
+                _mulitpleLocations = value; 
+                _props[PropertyName.MultipleLocations].IsSet = true;
+            }
+        }
+            
+    }
+    public DateTimeOffset Start {
+        get { return _start; }
+        set { _start = value; _props[PropertyName.Start].IsSet = true; }
+    }
+    public DateTimeOffset End {
+        get { return _end; }
+        set { _end = value; _props[PropertyName.End].IsSet = true; }
+    }
+
+    public class Property
+    {
+        public PropertyName Name { get; init; }
+        public bool IsSet = false;
+        public bool IsRequired = false;
+        public bool IsValid => (IsRequired && IsSet) || !IsRequired;
+        public string DisplayName { get; set; }
+
+        public Property(PropertyName name)
+        {
+            Name = name;
+            DisplayName = name.ToString();
+        }
+    }
+
+    internal Dictionary<PropertyName, Property> _props { get; init; }
+
+    public CarbonAwareParameters(Dictionary<string, string>? displayNameMap = null)
+    {
+        _props = InitProperties();
+        if (displayNameMap is not null) { _props.UpdateDisplayNames(displayNameMap); }
+    }
+
+    /// <summary>
+    /// Validates the properties and relationships between properties. This includes
+    ///  - Check that required properties are set
+    ///  - Check that specified relationships between properties (like start < end) are true
+    /// </summary>
+    public void Validate()
+    {
+        var errors = new Dictionary<string, List<string>>();
+
+        var multipleLocations = _props[PropertyName.MultipleLocations];
+        var start = _props[PropertyName.Start];
+        var end = _props[PropertyName.End];
+
+        // Validate Properties
+        if (!multipleLocations.IsValid) { errors.AppendValue(multipleLocations.DisplayName, $"{multipleLocations.DisplayName} is not set"); }
+        if (!start.IsValid) { errors.AppendValue(start.DisplayName, $"{start.DisplayName} is not set"); }
+        if (!end.IsValid) { errors.AppendValue(end.DisplayName, $"{end.DisplayName} is not set"); }
+
+        // Want to throw any property validation errors before checking any relationships between properties
+        CheckErrors(errors); 
+
+        // Validate Relationships
+        if (Start >= End) { errors.AppendValue(start.DisplayName, $"{start.DisplayName} must be before {end.DisplayName}"); }
+
```

Comment:
Instead of calling the CheckErrors() method twice, can we do the relationship validation if errors is not empty and call CheckErrors() once?
`if(!errors.Keys.Any() 
{
  if (Start >= End) { errors.AppendValue(start.DisplayName, $"{start.DisplayName} must be before {end.DisplayName}"); }
}`

Reply:
updated

Reply:
ok got it.. So coming back to my original comment, would this approach stagger the display of errors to the user rather than showing all the errors at once? I can ping you offline to understand this better.

Reply:
No not stagger. All the property validation errors will be packaged together into the ArgumentException. The reason for splitting is because we don't want to do any relationship validation if the property validation fails, because they we'd need to do a bunch of casing to make sure the relationship, we're checking had valid properties. So instead, we chose to throw first if any property validation happens and only do relationship validation if it passes

Reply:
@JenMadiedo and I chatted out-of-band and decided that `AssertNoErrors(errors)` would better communicate what this method does.  Thanks for the callout @pritipath !